### PR TITLE
[Mesh] initialize mesh dialects and passes

### PR DESCRIPTION
**Pull request by yaochengji**
- Mesh Dialects:
  - ops: mesh.cluster, mesh.idx, mesh.size, mesh.annotate, mesh.all_reduce, mesh.all_gather, mesh.reduce_scatter, mesh.all_to_all, mesh.local_split, 
  - attrs: MeshSharding
  - types: ShardingIteratorType
- Interfaces:
  - InferDTensorInterface
  - ShardingInterface
  - ShardingInterface Impl: mhlo.dot, mhlo.dot_general, mhlo.constant, elementwise ops
- Passes:
  - sharding propagation
  - sharding partition

### CODE REVIEW

File: compiler/test/Dialect/Mesh/ops.mlir [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// RUN: byteir-opt %s | FileCheck %s
+
+mesh.cluster @mesh0(rank = 3, dim_sizes = [4, 8, 12])
```
Comment (liwenchangbdbz) [inline compiler/test/Dialect/Mesh/ops.mlir:3]: is mesh.cluster per module?

Comment (yaochengji) [reply compiler/test/Dialect/Mesh/ops.mlir:3]: mesh.cluster is defined inside a module op and there could be multiple mesh.cluster op per module.
It should be useful for pipeline parallel, maybe I should supplement some design doc for it.

Comment (liwenchangbdbz) [reply compiler/test/Dialect/Mesh/ops.mlir:3]: or one example for pp or multiple mesh.cluster's will be good.

**Comment by yaochengji** [comment]
Closed due to most of the code will be contributed to MLIR repo.

**Pull request closed without merging**